lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , Role (..)
+    , SoftDerivation (..)
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..)
+    , VerificationKeyPool (..)
+    , defaultAddressPoolGap
+    , getAddressPoolGap
+    , mkUnboundedAddressPoolGap
+    , mkVerificationKeyPool
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( AddressState (..) )
+import Control.Monad
+    ( foldM )
+import Data.Functor.Identity
+    ( Identity (..) )
+
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) = compare sh1 sh2
+
+isShared
+    :: (k ~ ShelleyKey, SoftDerivation k)
+    => Script
+    -> SeqState n k
+    -> ([k 'ScriptK XPub], SeqState n k)
+isShared script (SeqState !s1 !s2 !ixs !rpk !prefix !scripts !s3) =
+    let verKeysInScript = retrieveAllVerKeyHashes script
+        (VerificationKeyPool accXPub currentGap verKeyMap) = s3
+        projectKey (ShelleyKey k) = ShelleyKey k
+        toVerKey = deriveAddressPublicKey accXPub MultisigScript
+        projectIndex = toEnum . fromInteger . toInteger . getIndex @'Soft
```

Comment:
I don't get this triple transformation :no_mouth:  ?

Reply:
it is just coerce :sweat_smile:  -> https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/b93c3fbe509ab5a5aa97a78f99f7207a5e2567f3